### PR TITLE
Finished Task 1 and Task 2

### PR DESCRIPTION
**Pull request by zinkoviktor**
**Task 1:**
1. In BusinessLayer project:
1.1. Crate folder Manager
1.2. In Manager folder create Interface IProductManager, the interface includes the declaration of GetAll() method
1.3. In Manager folder create class ProductManager which inherits from ProductManager
1.4. In ProductManager class implement GetAll() method which contains stub with list of three created and filled some data ProductModels entities and returns that list
2. Create ServiceLayer project:
2.1. In ServiceLayer project create DataTransferObjects folder
2.2. In DataTransferObjects folder create ProductDTO class with next fields:
    - Name = string type;
    - Description = string type;
    - Categories = string type;
    - Price = decimal type;
    - AvailableCount = int type;
2.3. In ServiceLayer project create Converters folder
2.4. In Converters folder create ProductConverter class with next methods:
    - ConverToDTO method with implemented logic to convert object of ProductModel type to object of ProductDTO type and the method returns converted object
    - ConverToModel method with implemented logic to convert object of ProductDTO type to object of ProductModel type and the method returns converted object

**Task 2:**
1. Add new project from empty ASP.NET Core Web Application template, project should contains next files:
1.1. appsettings.json
1.2. Program.cs
1.3. Startup.cs
2. Create Controllers folder
3. In Controllers folder create ProductController class, inherits its from ControllerBase class. Another case, add controller class from template in Visual Studio and add [Route("api/[controller]")] attribute to the class
4. In ProductsController class create ActionResult<IEnumerable<ProductDTO>> type method Get() and add [HttpGet] attribute to the method
5. In Get() method create instance of ProductManager class and instance of ProductConverter class and return Ok(...) method with converted ProductConverter Product instances which returning get from GetAll() method from ProductManager class
6. In ConfigureServices method from Startup class add services.AddControllers() method, and in Configure method change 'endpoints' in app.UseEndpoints method to invoking MapControllers() method

### CODE REVIEW

File: NetCoreRestApi/BusinessLayer/Manager/IProductManager.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using BusinessLayer.Models;
+using System.Collections.Generic;
+
+namespace BusinessLayer.Manager
+{
+    interface IProductManager
+    {
+        public IEnumerable<ProductModel> GetAll();
```
Comment (mdemii) [inline NetCoreRestApi/BusinessLayer/Manager/IProductManager.cs]: Remove public all methods are public in interface

File: NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+﻿using BusinessLayer.Models;
+using System.Collections.Generic;
+
+namespace BusinessLayer.Manager
+{
+    public class ProductManager : IProductManager
+    {
+        public IEnumerable<ProductModel> GetAll()
+        {
+            List<ProductModel> productModels = new List<ProductModel>();
```
Comment (mdemii) [inline NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs]: Use var if it is clear what type it is.

File: NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            productModels.Add(new StubProductModel("first product"));
+            productModels.Add(new StubProductModel("second product"));
+            productModels.Add(new StubProductModel("third product"));           
+            return productModels;
+        }
+        
+        private class StubProductModel : ProductModel
```
Comment (mdemii) [inline NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs]: One class per one file. Move this class into separated class

File: NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            internal StubProductModel(string name)
+            {
+                Name = name;
+                Description = "Description for " + name;
+                CategoryList = new List<CategoryModel>();
+                CategoryList.Add(new CategoryModel() { Name = "first category name", Description = "description" });
```
Comment (mdemii) [inline NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs]: Use short name like Category1, Category2, Category3 or real names like Food, Cars, Clothes e.t.c

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+using ServiceLayer.DataTransferObjects;
+using System.Collections.Generic;
+
+namespace ServiceLayer.Converters
+{
+    public class ProductConverter
+    {
+        public ProductDTO ConverToDTO(ProductModel productModel)
+        {            
+            if (productModel == null) return null;
```
Comment (mdemii) [inline NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: use { } for each single if block

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            productModel.CategoryList ??= new List<CategoryModel>();
```
Comment (mdemii) [inline NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs:15]: move initialization to ProductModel class

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            ProductDTO productDTO = new ProductDTO
```
Comment (mdemii) [inline NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: var

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                CategoryList = string.Join(", ", productModel.CategoryList.ConvertAll(x => x.Name)),
+                Price = productModel.Price,
+                AvailableCount = productModel.AvailableCount
+            };
+            return productDTO;
+        }       
+
+        public ProductModel ConverToModel(ProductDTO productDTO)
+        {
+            if (productDTO == null) return null;
```
Comment (mdemii) [inline NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: use {}

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            productDTO.CategoryList ??= "";
+            var categoryList = productDTO.CategoryList.Split(",");
+            ProductModel productModel = new ProductModel
+            {
+                Name = productDTO.Name,
+                Description = productDTO.Description,
+                Price = productDTO.Price,
+                AvailableCount = productDTO.AvailableCount,
+                CategoryList = new List<CategoryModel>()
```
Comment (mdemii) [inline NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: it is better to move this initialization to property

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            };
+            if (categoryList[0].Length > 0)
```
Comment (mdemii) [inline NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: this check is redundant. remove it

File: NetCoreRestApi/ServiceLayer/DataTransferObjects/ProductDTO.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+﻿
+namespace ServiceLayer.DataTransferObjects
+{
+    public class ProductDTO
+    {
+        public string Name { get; set; }
+        public string Description { get; set; }
+        public string CategoryList { get; set; }
```
Comment (mdemii) [inline NetCoreRestApi/ServiceLayer/DataTransferObjects/ProductDTO.cs]: It is not list. it is string. Rename to Categories

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+namespace WebAPI.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class ProductsController : ControllerBase
+    {
+        [HttpGet]
+        public ActionResult<IEnumerable<ProductDTO>> Get()
+        {
+            ProductManager productManager = new ProductManager();
```
Comment (mdemii) [inline NetCoreRestApi/WebAPI/Controllers/ProductsController.cs]: wrong initialization of productManager. it should be initialized over DI . Put IProductManager interface as private field. Inject interface into controller constuctor.

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+            var products = productManager.GetAll();
+            ProductConverter productConverter = new ProductConverter();
```
Comment (mdemii) [inline NetCoreRestApi/WebAPI/Controllers/ProductsController.cs]: Also inject product converter

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+            if (products == null) return NoContent();
```
Comment (mdemii) [inline NetCoreRestApi/WebAPI/Controllers/ProductsController.cs]: brackets

File: NetCoreRestApi/WebAPI/Startup.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace WebAPI
+{
+    public class Startup
+    {
+        // This method gets called by the runtime. Use this method to add services to the container.
+        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddControllers();
```
Comment (mdemii) [inline NetCoreRestApi/WebAPI/Startup.cs:16]: where are interfaces for ProductManager, Converters?  DI?

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs (hunk 6 above)
Comment (zinkoviktor) [reply NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs:15]: Initialization needs only if product Model.Category List is null, for prevention throwing null reference exeption

File: NetCoreRestApi/WebAPI/Startup.cs (hunk 15 above)
Comment (zinkoviktor) [reply NetCoreRestApi/WebAPI/Startup.cs:16]: Fixed below, Is it right?

File: NetCoreRestApi/WebAPI/Startup.cs [hunk 16]
```diff
@@ -12,6 +14,8 @@ public class Startup
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddControllers();
+            services.AddSingleton<IProductManager, ProductManager>();
+            services.AddSingleton<IProductConverter, ProductConverter>();
```
Comment (mdemii) [inline NetCoreRestApi/WebAPI/Startup.cs]: Use AddTransient

Comment (zinkoviktor) [reply NetCoreRestApi/WebAPI/Startup.cs]: Done

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs (hunk 14 above)
Comment (zinkoviktor) [reply NetCoreRestApi/WebAPI/Controllers/ProductsController.cs]: Done

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs (hunk 13 above)
Comment (zinkoviktor) [reply NetCoreRestApi/WebAPI/Controllers/ProductsController.cs]: Done

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs (hunk 12 above)
Comment (zinkoviktor) [reply NetCoreRestApi/WebAPI/Controllers/ProductsController.cs]: Done

File: NetCoreRestApi/ServiceLayer/DataTransferObjects/ProductDTO.cs (hunk 11 above)
Comment (zinkoviktor) [reply NetCoreRestApi/ServiceLayer/DataTransferObjects/ProductDTO.cs]: Done

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs (hunk 10 above)
Comment (zinkoviktor) [reply NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: Done

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs (hunk 9 above)
Comment (zinkoviktor) [reply NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: Done

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs (hunk 8 above)
Comment (zinkoviktor) [reply NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: Done

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs (hunk 7 above)
Comment (zinkoviktor) [reply NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: Done

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs (hunk 6 above)
Comment (zinkoviktor) [reply NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs:15]: Done

File: NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs (hunk 5 above)
Comment (zinkoviktor) [reply NetCoreRestApi/ServiceLayer/Converters/ProductConverter.cs]: Done

File: NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs (hunk 4 above)
Comment (zinkoviktor) [reply NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs]: Done

File: NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs (hunk 3 above)
Comment (zinkoviktor) [reply NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs]: Done

File: NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs (hunk 2 above)
Comment (zinkoviktor) [reply NetCoreRestApi/BusinessLayer/Manager/ProductManager.cs]: Done

File: NetCoreRestApi/BusinessLayer/Manager/IProductManager.cs (hunk 1 above)
Comment (zinkoviktor) [reply NetCoreRestApi/BusinessLayer/Manager/IProductManager.cs]: Done

**Review by mdemii — APPROVED** [review]

**Pull request merged**